### ISSUE TITLE
UnixHttpClientTestCase.java:50-52: Implement a unix...

### ISSUE THREAD

**Issue by 0pdd** [amihaiemil/docker-java-api#41]
The puzzle `38-ae2c79c8` from #38 has to be resolved:

https://github.com/amihaiemil/docker-java-api/blob/23aa92a82f3ee0f8fa258716bfee7fce889872a4/src/test/java/com/amihaiemil/docker/UnixHttpClientTestCase.java#L50-L52

The puzzle was created by George Aristy on 13-Mar-18. 

Estimate: 30 minutes, role: DEV.

If you have any technical questions, don't ask me, submit new tickets instead. The task will be "done" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) please, pay attention to this issue

**Comment by 0crat** [comment]
Job [#41](https://github.com/amihaiemil/docker-java-api/issues/41) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
The job [#41](https://github.com/amihaiemil/docker-java-api/issues/41) assigned to @llorllale[/z](https://www.0crat.com/u/llorllale), here is [why](https://www.0crat.com/footprint/G6LPQQV2P/881371023); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html)

**Comment by amihaiemil** [comment]
@llorllale There are some examples in the unix library we're using: https://github.com/jnr/jnr-unixsocket/tree/master/src/test/java/jnr/unixsocket/example 

I think something similar to that ``UnixServer.java`` would be enough :-?

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by llorllale** [comment]
@amihaiemil that example sure is a huge help but I'm still not sure how to tackle this. Do you know if that while loop in `UnixServer.main` needs to be run in a separate thread?

**Comment by llorllale** [comment]
@amihaiemil nevermind, I think it does. 

Beyond implementing a server: what functions do you want it to do to aid us in tests? Should it accept an `AssertRequest` in order to validate the request sent? Should it also accept `Response` objects?

**Comment by llorllale** [comment]
@amihaiemil here's a draft:

```java
/**
 * A simple UNIX socket server.
 * @author George Aristy (george.aristy@gmail.com)
 * @version $Id$
 * @since 0.0.1
 */
public final class UnixServer implements Closeable {
    private final Path socketFile;
    private final UnixServerSocketChannel channel;
    private final Selector selector;
    private final ExecutorService threadPool;
  
    /**
     * Ctor.
     * <p>
     * Eagerly opens a unix file socket created in a temp directory.
     * @throws IOException If an I/O error occurs.
     */
    public UnixServer() throws IOException {
        this.socketFile = Files.createTempFile("", "");
        this.channel = UnixServerSocketChannel.open();
        this.channel.configureBlocking(false);
        this.channel.socket().bind(new UnixSocketAddress(this.socketFile.toFile()));
        this.selector = NativeSelectorProvider.getInstance().openSelector();
        this.channel.register(this.selector, SelectionKey.OP_READ);
        this.threadPool = Executors.newSingleThreadExecutor();
        this.threadPool.submit(new Service(this.selector));
    }
  
    /**
     * The UNIX socket file.
     * @return The unix socket file opened by this server.
     */
    public Path socketFile() {
        return this.socketFile;
    }
  
    @Override
    public void close() throws IOException {
        this.threadPool.shutdownNow();
        this.selector.close();
        this.channel.close();
        Files.delete(this.socketFile);
    }
  
    /**
     * Background thread that handles requests to socket.
     */
    private static final class Service implements Callable<Void> {
        private final Selector selector;
  
        /**
         * Ctor.
         * @param channel 
         */
        private Service(Selector selector) {
            this.selector = selector;
        }
  
        @Override
        public Void call() throws IOException {
            while (this.selector.isOpen()) {
                this.selector.select();
                final Iterator<SelectionKey> iter = this.selector.selectedKeys()
                    .iterator();
                while (iter.hasNext()) {
                    final SelectionKey key = iter.next();
                    iter.remove();
                    final UnixSocketChannel channel
                        = (UnixSocketChannel) key.channel();
                    channel.configureBlocking(false);
                    final ByteBuffer buffer = ByteBuffer.wrap("".getBytes()); //WHAT SHOULD BE OUR RESPONSE?
                    channel.write(buffer);
                }
            }
            return null;
        }
    }
}
```

**Comment by amihaiemil** [comment]
@llorllale I'll have a closer look tomorrow... meanwhile, feel free to register impediments on your tasks if they are too many. After all, it's only us still, so it's not like 0crat can assign them to someone else.

**Comment by llorllale** [comment]
@0crat waiting while we figure this one out...

**Comment by 0crat** [comment]
> @0crat waiting while we figure this one out... ([here](https://github.com/amihaiemil/docker-java-api/issues/41#issuecomment-378395564))

@llorllale The impediment for [#41](https://github.com/amihaiemil/docker-java-api/issues/41) was registered successfully by @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by amihaiemil** [comment]
@llorllale Looks ok. I only think it should also accept a Response and write it back to the user. We don't need to assert the request itself since that is already tested in the current tests.

But, let's close this, you already spent the time, I think. And you can open a new Issue to continue implementation there :)

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) please, pay attention to this issue

**Comment by 0crat** [comment]
@0pdd[/z](https://www.0crat.com/u/0pdd) this project will fix the problem faster if you donate a few dollars to it; just [click here](https://www.0crat.com/contrib/G6LPQQV2P) and pay via Stripe, it's very fast, convenient and appreciated; thanks a lot!

**Comment by amihaiemil** [comment]
@llorllale ugh, bummer, I forgot it has a puzzle in the code. Can you make a PR just with removing the puzzle and adding a new one for continuing?

**Comment by 0crat** [comment]
GitHub issue is still open, won't close

**Comment by 0pdd** [comment]
The puzzle `38-ae2c79c8` has disappeared from the source code, that's why I closed this issue.

**Comment by 0pdd** [comment]
@0pdd the puzzle [#79](https://github.com/amihaiemil/docker-java-api/issues/79) is still not solved.

**Comment by 0crat** [comment]
Order was finished: +30 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by 0crat** [comment]
The job [#41](https://github.com/amihaiemil/docker-java-api/issues/41) is now out of scope

**Comment by 0pdd** [comment]
@0pdd the puzzle [#87](https://github.com/amihaiemil/docker-java-api/issues/87) is still not solved; solved: [#79](https://github.com/amihaiemil/docker-java-api/issues/79).

**Comment by 0pdd** [comment]
@0pdd the puzzle [#87](https://github.com/amihaiemil/docker-java-api/issues/87) is still not solved; solved: [#10](https://github.com/amihaiemil/docker-java-api/issues/10), [#79](https://github.com/amihaiemil/docker-java-api/issues/79).
